docs/proposals/20220209-machinepool-machines.md

```diff
@@ -0,0 +1,195 @@
+
+### Goals
+
+- Enable CAPI MachinePool to create Machines to represent infrastructure provider MachinePool Machines
+- Ensure that Machines in MachinePools can participate in the same behaviors as MachineDeployment Machines (node drain, machine health checks, retired node deletion)
+- Enhance `clusterctl describe` to display MachinePool Machines
+- Allow CAPI MachinePool to include deployment strategies mirroring those of MachineDeployments (future work)
+
+### Non-Goals/Future Work
+
+- Require infrastructure providers to implement MachinePool Machines. This is intended to be an "opt-in" feature, since there are existing MachinePool implementations that cannot support it. For example, `AWSManagedMachinePool` does not allow for deleting individual instances.
+- Implement Cluster Autoscaler MachinePool integration (a separate task, although we have a working POC)
+- Move the MachinePools implementation into a `exp/machinepools` package. The current code location makes its separation from `exp/addons` unclear and this would be an improvement.
+- Improve bootstrap token handling for MachinePools. It would be preferable to have a single, short-lived token for each Machine instead of the current shared, long-lived token.
+## Proposal
+
+To create MachinePool Machines, a MachinePool in CAPI needs information about the instances or replicas associated with the provider's implementation of the MachinePool. This information is attached to the provider's MachinePool infrastructure resource as new, optional status fields `InfrastructureMachineSelector`. This field may be populated by the infrastructure provider.
+
+```golang
+// FooMachinePoolStatus defines the observed state of FooMachinePool.
+type FooMachinePoolStatus struct {
+	// InfrastructureMachineSelector is a label query over FooMachinePoolMachines.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+	// +optional
+	InfrastructureMachineSelector metav1.LabelSelector `json:"infrastructureMachineSelector,omitempty"`
+}
+```
+
+If this field is populated, CAPI's MachinePool controller will query for the provider-specific infrastructure resources. That query uses the Selector field, the API version of the \<Provider\>MachinePool, and the Kind of the \<Provider\>MachinePool with "Machine" appended to it. For example, if the \<Provider\>MachinePool is of Kind "FooMachinePool", the query will be of Kind "FooMachinePoolMachine".
+
+Once found, CAPI will create and connect MachinePool Machines to each resource. A MachinePool Machine is implemented as a Cluster API Machine that is owned by a MachinePool. If the field is empty, no MachinePool Machines will be created, and the overall behavior is the same as before this proposal was implemented.
```

Comment:
If the new selector is not set, the components such as cluster autoscaler wont be able to scale up and scale down is it? Wont that lead to confusion?